### ISSUE TITLE
Interface merge of import/local should error when the import is re-exported with `export *`

### ISSUE THREAD

**Issue by unional** [Microsoft/TypeScript#9532]
**TypeScript Version:**  (2.0.0-dev.20160705)

**Code**

```ts
// foo.ts
declare module 'A' {
  export interface Foo<S> {
    (): void;
  }
}
declare module '~B' {
  import { Foo as Roo } from 'A';
  interface Foo extends Roo<any> {
  }
}
declare module 'B' {
export * from '~B';
}

// boo.ts
declare module 'B' {
  interface Foo {
     (message: string): any;
  }
}

export const SOME = 1;

// go.ts
import { Foo } from 'B';

let x: Foo;
x(); // <-- only have (message: string): any

```

**Expected behavior:**
`x();` have overloads of `(message: string): any` and `(): void`

**Actual behavior:** 
`x();` only have `(message: string): any`

If I change `boo.ts` to:
```ts
// boo.ts
import { Foo as Soo } from 'A';
declare module 'B' {
  interface Foo extends Soo<any> { // <-- i.e. matching original extends
     (message: string): any;
  }
}

export const SOME = 1;
```

Then it works correctly.

Also, if I reference `~B` instead of `B` (i.e. without re-export), it works correctly too:

```ts
// boo.ts
declare module '~B' {
  interface Foo {
     (message: string): any;
  }
}

export const SOME = 1;

// go.ts
import { Foo } from '~B';

let x: Foo;
x()
```

**Comment by unional** [comment]
cc @blakeembrey to see if it rings any bell on similar issues.

**Comment by unional** [comment]
Expanding this to cover one more issue:

```ts
// foo.ts
declare module 'A' {
  export interface Foo<S> {
    (): void;
  }

  export class SomeClass<S> {
    foo: Foo<S>;
  }
}
declare module '~B' {
  import { Foo as Roo, SomeClass } from 'A';
  export { SomeClass }
  interface Foo extends Roo<any> {
  }
}
declare module 'B' {
export * from '~B';
}

// boo.ts
import { Foo as Soo } from 'A';

declare module 'B' {
  interface Foo extends Soo<any> { // <-- i.e. matching original extends
    (message: string): any;
  }
}

export const SOME = 1;

// go.ts
import { Foo, SomeClass } from 'B';

let x: Foo;
x(); // <-- has overloads

let y: SomeClass<any>;
y.foo(); // <-- no overloads
```

**Comment by unional** [comment]
This issue is more common than I thought.

It would happen if I wrote a module distributed through npm this way:

```ts
// index.ts
export * form './a';
export * from './b';

// a.ts
...

// b.ts
...
```

Which is a pretty common pattern.

**Comment by MI-DEV-02** [comment]
Maybe related: https://github.com/Microsoft/TypeScript/issues/9127

**Comment by paleo229** [comment]
Maybe related: https://github.com/Microsoft/TypeScript/issues/9127

**Comment by sandersn** [comment]
This is basically a duplicate of #8140. You can't augment things that are imported from a module, even types. But the bad error you found is different than #8140. `export * from "~B"` doesn't report the error, but `export { Foo } from "~B"` does. The compiler should give the same error for `export *` that it does for `export { Foo }`.

I'll give more detail in a subsequent comment. I'll include a modified repro and an explanation of why this fails to do what you want.

**Comment by unional** [comment]
> You can't augment things that are imported from a module, even types

Not sure if I understand. `boo.ts` is not importing module `B`, it just augment module `B`.

**Comment by sandersn** [comment]
Here's the explanation I promised. I assume that you are trying to augment the `Foo` declared in `~B` based on what you expect to happen. The `declare module 'B'` block in boo.ts is supposed to do this. Unfortunately, in foo.ts, `B` re-exports `Foo` from `~B`, so you can't -- once you import something, it will never merge:

```ts
declare module 'B' {
  export * from '~B'; // import and re-export `Foo`
}
```

TypeScript should error here. And it does -- if you explicitly say `export { Foo } from '~B'`. It incorrectly fails to error with `export * from '~B'`.

In boo.ts, you could `declare module '~B'` and augment `~B` instead of `B` instead. This works as long as all your modules are originally declared as ambient module declarations (that is, `declare module 'X'` blocks in a file consisting *only* of `declare module 'X'` blocks).

**Comment by sandersn** [comment]
I forgot to mention earlier that TypeScript skips the re-exported `Foo` from `B` that it should have errored on. That's why the base members are missing that you expect to see.

Here's the modified repro that uses properties:
```ts
// foo.ts
declare module 'A' {
  export interface Foo {
    foo: number; // diff 1: use property instead of call signature
  }
}
declare module '~B' {
  import { Foo as Roo } from 'A';
  interface Foo extends Roo {
      baz: boolean; // diff 2: add property
  }
}
declare module 'B' {
export * from '~B';
}

// boo.ts
declare module 'B' {
  interface Foo {
     bar: string; // diff 3: use property instead of call sig
  }
}
export const SOME = 1;

// go.ts
import { Foo } from 'B';

let x: Foo = {  // diff 4: initialise to object literal
    foo: 1,  // error: 'foo' does not exist in type 'Foo'
    bar: 'maybe', 
    baz: false, // error: 'baz' does not exist in type 'Foo'
};
```

The simplest fix is a one-liner. In boo.ts, augment `'~B'` instead. This allows `Foo` to merge *before* being exported by `B`:

```ts
// boo.ts
declare module '~B' {
  interface Foo {
    bar: string;
  }
}
```

I tried your fix to this modified repro but it still says that 'baz' does not exist in type 'Foo':

```ts
// boo.ts
import { Foo as Soo } from 'A';
declare module 'B' {
  interface Foo extends Soo {
     bar: string;
  }
}
export const SOME = 1;
```

The reason that this doesn't work is that the re-export is still skipped so `baz` is still missing (and typescript still fails to error). But since `Foo` now explicitly extends `Soo`, the compiler at least knows that `foo` is a member of `Foo`.

**Comment by unional** [comment]
Does that means module written like this:

```ts
// index.ts
export * form './a';
export * from './b';

// a.ts
...

// b.ts
...
```

cannot be augmented?

**Comment by unional** [comment]
Also, this could cause issue on `typings` if the re-export can't be merged.

**Comment by unional** [comment]
When you say "Interface merge of alias/local should error when the alias is re-exported with `export *`", do you mean it is the alias `Roo` in  `import { Foo as Roo } from 'A';` causing the issue here?

**Comment by sandersn** [comment]
Which module do you want to augment? `index.ts`? That's correct. You are not allowed to modify things that are imported from modules. Looking at existing libraries on DefinitelyTyped, it looks like people either use `/// <reference ...` or else put the whole core of their library into `index.d.ts`.

**Comment by sandersn** [comment]
No, it is the `export * from '~B'`. If you rewrite that line as `export { Foo } from '~B'`, you'll see the error "Export declaration conflicts with exported declaration of 'Foo': interface Foo, import Foo".

I think this could indeed be an issue for `typings`. I'll bring it to the attention of @RyanCavanaugh.

**Comment by sandersn** [comment]
The term "alias" is misleading. Internally, the compiler refers to any imported symbol as an alias, whether it actually uses the `as alias` syntax or not.

**Comment by unional** [comment]
> use /// <reference ... or else put the whole core of their library into index.d.ts

Do you mean `index.ts`? This limitation would definitely surprise many people.

**Comment by unional** [comment]
> see the error "Export declaration conflicts with exported declaration of 'Foo': interface Foo, import Foo"

In foo.ts, I do not see that error using 7/5 build
![image](https://cloud.githubusercontent.com/assets/3254987/16666365/26835c3c-443c-11e6-9523-9a403ad17a06.png)

**Comment by sandersn** [comment]
a `.d.ts` is just a TypeScript file that contains only declarations -- but projects like [Angular](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/angularjs/angular.d.ts) put a lot of declarations in their top-level declaration file, not just re-exports. I think you are right that this limitation will surprise people, however.

What tool fails to produce the error? I see it in VS Code and running `tsc` from the command line. I am using the code from this morning.

**Comment by unional** [comment]
> What tool fails to produce the error?

That's strange. I am using VSCode. It doesn't work for both past and today's release. With 7/5 TS nightly.

**Comment by unional** [comment]
Is this a technical limitation or there are use case for this limitation?
This will break all augmentation for `typings` I believe.  cc @blakeembrey

**Comment by blakeembrey** [comment]
Thanks @unional, sounds like it. I had assumed this was fixed, but I guess I reviewed it incorrectly. For reference, there's a whole chain of issues starting from https://github.com/typings/typings/issues/525#issuecomment-226062689. Also, being unable to augment any `export *` modules is a huge gotcha. It's fairly common to write TypeScript that exports from multiple files as the main file (E.g. https://github.com/blakeembrey/jaywalk/blob/master/src/types/index.ts or https://github.com/blakeembrey/node-scrappy/blob/master/src/index.ts).

**Comment by unional** [comment]
> I had assumed this was fixed, but I guess I reviewed it incorrectly

I looked it up and it fixed the export default and augmentation in ADDING interface: https://github.com/Microsoft/TypeScript/issues/8840

![image](https://files.gitter.im/typings/3XKX/blob)

But this is about augmenting existing interface.

**Comment by unional** [comment]
Also on the js front it is common: https://github.com/lodash/lodash/blob/es/lodash.js
and everything in angular: https://github.com/angular/angular/blob/master/modules/%40angular/http/index.ts
https://github.com/angular/angular/blob/master/modules/%40angular/core/index.ts
...

**Comment by sandersn** [comment]
It's a technical limitation. I don't know yet if it's by design (and if so, if there's an intended workaround) -- I'm waiting on @RyanCavanaugh to get back from vacation since he knows much more than I do about typings.

**Comment by sandersn** [comment]
Unfortunately, after some more discussion with the team, we decided that we can't support this pattern for now. We might revisit this decision in the future, but current DefinitelyTyped code has worked around this limitation so far, which may indicate that it's (1) easy to work around or (2) not that common.

**Comment by blakeembrey** [comment]
@sandersn Unless I've misunderstood something, isn't this extremely common? Basically any node module that's re-exporting from multiple files follows this pattern (which is most node modules, you can regularly find anything bigger than a single function util exporting multiple files in `index.js`). I believe even TypeScript would be authored in this format if it were originally written as ES6 modules.

DefinitelyTyped "works around it" because it mutates the global scope. The only other possible work-around in ordinary code would be to augment the original source file. But that means any file in a module is now part of the public API, which is frowned upon doing and quite fragile.

**Comment by sandersn** [comment]
@RyanCavanaugh Thoughts? You've worked with DefinitelyTyped a lot more than I have.

**Comment by unional** [comment]
@RyanCavanaugh what's your thought?

I have a private library written in TypeScript that hit this limitation.
And example is to imagine `redux` is written in TypeScript and `redux-thunk` and `redux-promise` needs to augment the `Dispatch` method.

This bug prevents that from happening.

My workaround is put all source code into `index.ts` in the "`redux`" package. Which definitely won't scale as the package evolves.

**Comment by unional** [comment]
Pinging @RyanCavanaugh

**Comment by unional** [comment]
Any update on this? We should reconsider this, not just for typings, for package written in TypeScript in general.
